### ISSUE TITLE
Lesson 1, decoder to linear batch position

### ISSUE THREAD

**Issue by marcpaga** [bentrevett/pytorch-seq2seq#164]
Dear @bentrevett,

In the decoder code from your first lesson the RNN of the decoder outputs in shape [timesteps, batch, hidden_dim]

```
output, (hidden, cell) = self.rnn(embedded, (hidden, cell))
#output = [seq len, batch size, hid dim * n directions]
```

This output is then directly fed into a linear layer for the predictions

```
prediction = self.fc_out(output.squeeze(0))
```

Wouldn't this give the timesteps in the first dimension? According to the [pytorch documentation](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html#torch.nn.Linear), a linear layer expects the input in shape [batch, timesteps, hidden_dim]. Therefore shouldn't there be a `output.permute(1, 0, 2)` before?

Perhaps I am missing something.
Marc

**Comment by bentrevett** [comment]
> Wouldn't this give the timesteps in the first dimension?

No, because we `squeeze` the `output` in the `self.fc_out(output.squeeze(0))` line.

The `squeeze` converts `output` from `[seq len, batch size, hid dim * n directions]` to `[batch size, hid dim * n directions]`. 

I also believe it wouldn't actually matter, as the `nn.Linear` layer is always applied just to the last dimension of the input across all other dimensions. So if we had a `nn.Linear` layer that went from 100 to 10 dimensions (`nn.Linear(100, 10)`) our input to it could be `[300, 200, 100]` and internally the `nn.Linear` layer would reshape to `[300 * 200, 100]` and then apply the weight and bias to the final dimension only to get `[300 * 200, 10]` and then reshape back to `[300, 200, 10]`. 

But again, in the decoder we're doing the `squeeze` there's no internal reshaping or anything, just thought it might be interesting to you.

**Comment by marcpaga** [comment]
I see, I was not understanding the `.squeeze(0)`, but I now realize that there is only one timepoint.

Regarding your explanation of `nn.Linear`, does that mean that internally it treats each timepoint as a "sample" in the batch, and this makes it so that each timepoint is independent? 

Thanks for your explanations and responding so quickly!

**Comment by bentrevett** [comment]
> Regarding your explanation of `nn.Linear`, does that mean that internally it treats each timepoint as a "sample" in the batch, and this makes it so that each timepoint is independent?

Yep, if you pass in a batch of features over time then each feature (across batches and time-steps) is treated completely independently.

**Comment by marcpaga** [comment]
Thanks a lot for your explanation, this has been really helpful!
